book/src/proposals/leader-duplicate-block-slashing.md

```diff
@@ -0,0 +1,70 @@
+# Leader Duplicate Block Slashing
+
+This design describes how the cluster slashes leaders that produce
+duplicate blocks.
+
+Leaders that produce multiple blocks for the same slot increase the
+number of potential forks that the cluster has to resolve.
+
+## Overview
+
+1. Slashing Condition: If you have voted on version `S1` of a slot, it is a slashing condition to vote on any descendant of another version `S2` of the slot, unless that descendant of version `S2` has seen supermajority (> 66%) of the stake voting on it (See section `Slashing Condition` below for how this proof is to be generated).
+2. If you see two blockhashes for the same block, and you are less than 2^THRESHOLD lockout on that block, drop it. Flag the block as "duplicated".
+3. Only repair a version of the "duplicated" block if you see a `Repair Proof` as outlined in the `Repair Proof` section below
+
+## Primitives
+
+For a bank `B`, let `A` be the latest ancestor of `B` that has gotten supermajority > 66% votes. 
+
+1) Define the function `Confirmed` to be `Confirmed(B) = A`. 
+2) Define `parent(B)` to be the parent bank of `B`.
+
+The `bank_hash` of each bank `B` is now augmented to be: 
+
+`hash(Confirmed(A).slot, state_hash(B), bank_hash(parent(B)), parent(B).slot)`.
+
+## Repair Proof
+
+Let `S'` be some version of slot `S` that a validator has already detected a
+duplicate version of and dropped and marked as "duplicate"
+(step 2 of the `Overview` section above).  A validator will only repair 
+`S'` if it's provided a proof that `S'` has been confirmed 
+(been voted on by greater than 66% of the stake). 
+
+Call such a proof `RepairProof(S')`.
+
+This proof consists of a supermajority of validator's votes, where each vote `V`:
+
+* Contains a bank hash `H(B)` for some bank `B` such that `Confirmed(B).slot < S'.slot` (if a supermajority confirmed `S'`, then such a set must exist because there must have been some initial set of validators that voted on `S'` before it was confrmed). This can be confirmed for each vote if the prover provides `state_hash(B)`, `Confirmed(B).slot`, `bank_hash(parent(B))`, `parent(B).slot` such that `H(B) == hash(Confirmed(B).slot, state_hash(B), bank_hash(parent(B)), parent(B).slot)`.
+
+* Show that `V.slot` is descended from `S'` by proving the bank hash `H(B)` can be derived from the bank hash and slot of `S'`. This is done by providing the trail of bank hashes and parent slots to 
```

Comment:
How does a validator receiving the proof determine the weights of the votes?

Reply:
I was thinking they just use the epoch stake weights?

Reply:
I was thinking they just use the epoch stake weights?

Reply:
I was thinking we just use the epoch stakes?